crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+/// classDef pending fill:#FFCA3A
+/// classDef sidechain fill:#FF595E
+/// ```
+///
+///
+/// main functions:
+/// * insert_block: insert block inside tree. Execute it and save it to database.
+/// * finalize_block: Flush chain that joins to finalized block.
+/// * make_canonical: Check if we have the hash of block that we want to finalize and commit it to
+///   db. If we dont have the block pipeline syncing should start to fetch the blocks from p2p.
+/// *
+/// Do reorg if needed
+
+pub struct BlockchainTree<DB: Database, C: Consensus, EF: ExecutorFactory> {
+    /// chains and present data
+    pub chains: HashMap<ChainId, Chain>,
+    /// Static chain id generator
+    pub chain_id_generator: u64,
+    /// Indices to block and their connection.
+    pub block_indices: BlockIndices,
+    /// Number of block after finalized block that we are storing. It should be more then
+    /// finalization window
+    pub num_of_side_chain_max_size: u64,
+    /// Finalization windows. Number of blocks that can be reorged
+    pub finalization_window: u64,
+    /// Externals
+    pub externals: Externals<DB, C, EF>,
+}
+
+/// Container for external abstractions.
+pub struct Externals<DB: Database, C: Consensus, EF: ExecutorFactory> {
+    /// Save sidechain, do reorgs and push new block to canonical chain that is inside db.
+    pub db: DB,
+    /// Consensus checks
+    pub consensus: C,
+    /// Create executor to execute blocks.
+    pub executor_factory: EF,
+    /// Chain spec
+    pub chain_spec: Arc<ChainSpec>,
+}
+
+impl<DB: Database, C: Consensus, EF: ExecutorFactory> Externals<DB, C, EF> {
+    /// Return sharable database helper structure.
+    pub fn sharable_db(&self) -> ShareableDatabase<&DB> {
+        ShareableDatabase::new(&self.db, self.chain_spec.clone())
+    }
+}
+
+/// Helper structure that wraps chains and indices to search for block hash accross the chains.
+pub struct BlockHashes<'a> {
+    /// Chains
+    pub chains: &'a mut HashMap<ChainId, Chain>,
+    /// Indices
+    pub indices: &'a BlockIndices,
+}
+
+impl<DB: Database, C: Consensus, EF: ExecutorFactory> BlockchainTree<DB, C, EF> {
+    /// New blockchain tree
+    pub fn new(
+        externals: Externals<DB, C, EF>,
+        finalization_window: u64,
+        num_of_side_chain_max_size: u64,
+        num_of_additional_canonical_block_hashes: u64,
+    ) -> Result<Self, Error> {
+        if finalization_window > num_of_side_chain_max_size {
+            panic!("Side chain size should be more then finalization window");
+        }
+
+        let last_canonical_hashes = externals
+            .db
+            .tx()?
+            .cursor_read::<tables::CanonicalHeaders>()?
+            .walk_back(None)?
+            .take((finalization_window + num_of_additional_canonical_block_hashes) as usize)
+            .collect::<Result<Vec<(BlockNumber, BlockHash)>, _>>()?;
+
+        // TODO(rakita) save last finalized block inside database but for now just take
```

Comment:
Is this still a todo?

Reply:
Yes, it would require putting last known finalise block number somewhere in database. It is low priority and can be added later, we now just assume `last_tip-64`

Reply:
issue here: https://github.com/paradigmxyz/reth/issues/1712